docs: expand documentation about testing

wplapper opened the pull request:
Add "Writing tests" to the docs.

<!--
Thank you very much for contributing code or documentation to restic! Please
fill out the following questions to make it easier for us to review your
changes.
-->

What does this PR change? What problem does it solve?
-----------------------------------------------------

<!--
Describe the changes and their purpose here, as detailed as needed.
-->
This PR adds a few paragraphs about adding tests, and where to look for basic patterns. 

Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------

<!--
Link issues and relevant forum posts here.

If this PR resolves an issue on GitHub, use "Closes #1234" so that the issue
is closed automatically when this PR is merged.
-->
https://github.com/restic/restic/issues/5252
Closes #5252 

Checklist
---------

<!--
You do not need to check all the boxes below all at once. Feel free to take
your time and add more commits. If you're done and ready for review, please
check the last box. Enable a checkbox by replacing [ ] with [x].

Please always follow these steps:
- Read the [contribution guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches).
- Enable [maintainer edits](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- Run `gofmt` on the code in all commits.
- Format all commit messages in the same style as [the other commits in the repository](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits).
-->

~~- [ ] I have added tests for all code changes.~~
- [X] I have added documentation for relevant changes (in the manual).
- [X] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [ ] I'm done! This pull request is ready for review.

wplapper committed:
Adding to the documetation about testing.

Add "Writing tests" to the docs.

wplapper committed:
docs testing machinery

Added PR number

wplapper commented:
I have decided to revert this PR to closed status to ease the pressure on the 
maintainers of restic. Once the number of PRs has dropped considerably, I might
moved this PR back to the queue to be reviewed. In case you consider this PR
important, please contact me via the usual channels. Thank you!

MichaelEischer commented on changelog/unreleased/issue-5252:
```diff
@@ -0,0 +1,7 @@
+Enhancement: Add a documentation section about "Writing tests" for restic
```
Not necessary for a developer only change.

MichaelEischer commented on doc/090_participating.rst:
```diff
@@ -105,6 +105,77 @@ project <https://dave.cheney.net/2016/03/12/suggestions-for-contributing-to-an-o
+ rtest.Assert(t, a == b, "msg")     checks for a condition to be true
+ rtest.OK(t, err)                   expects err to be ``nil``, otherwise fail test
+ rtest.OKs(t, errs)                 expects a slice of errs to be ``nil``
+ testListSnapshots(t, env.gopts, n) expects exactly <n> snapshots in the repository
```
rtest and testListSnapshots should be explained separately.

MichaelEischer commented on doc/090_participating.rst:
```diff
@@ -105,6 +105,77 @@ project <https://dave.cheney.net/2016/03/12/suggestions-for-contributing-to-an-o
+
+If you need to capture output from a command, look out for examples with ``withCaptureStdout`` in
+``cmd/restic/*_test.go``. Basicly, stdout gets attached to a ``bytes.Buffer``.
+In addition, you can use the global option ``gopts.JSON: true`` to generate JSON output
```
```suggestion
In addition, you can set the option ``gopts.JSON: true`` to generate JSON output
```

Even though gopts is called global options, it's not global in test cases.

MichaelEischer commented on doc/090_participating.rst:
```diff
@@ -105,6 +105,77 @@ project <https://dave.cheney.net/2016/03/12/suggestions-for-contributing-to-an-o
+ rtest.Assert(t, len(lines) == 2, ...)         // make sure that you have correct output
+
+When it comes to writing tests in the ``internal`` directory tree, things can become
+a bit more tricky, since you haven't necessarily got the full support of an opened repository
```
That's the whole point and in many cases is perfectly fine. See https://github.com/restic/restic/pull/5346/changes#r2751469634 regarding what to use instead.

MichaelEischer reviewed (changes requested):
The current description focuses too much on the integration tests. To make it clearer, please properly untangle the unit test and the integration test parts.

wplapper committed:
testing documentation - headline

Changed headline to Change: ...

wplapper committed:
documentation for tuning

fixed wrong link, missing leading "_"

wplapper committed:
doc - writing tests - apply review

Reworked the chapter about writing tests completely, split chapter into
two sub-chapters: unit tests and integration tests.

MichaelEischer commented on doc/090_participating.rst:
```diff
@@ -105,6 +105,129 @@ project <https://dave.cheney.net/2016/03/12/suggestions-for-contributing-to-an-o
+
+The packages in ``internal/...`` often provide ``Test*(...)`` functions and structs or
+have a dedicated test package like ``internal/backend/test``.
+A good starting points are also the testing.go files that exist in several places.
```
```suggestion
A good starting points are also the `testing.go` files that exist in several places.
```

MichaelEischer commented on doc/090_participating.rst:
```diff
@@ -105,6 +105,129 @@ project <https://dave.cheney.net/2016/03/12/suggestions-for-contributing-to-an-o
+``checker.TestCheckRepo()`` can be used to verify the repository integrity.
+
+In general, in memory operation should be preferred over creating temporary files.
+If necessary, store them in ``t.TempDir()``. However, in most cases only a
```
```suggestion
If necessary, temporary files can be stored in ``t.TempDir()``. However, in most cases only a
```

MichaelEischer commented on doc/090_participating.rst:
```diff
@@ -105,6 +105,129 @@ project <https://dave.cheney.net/2016/03/12/suggestions-for-contributing-to-an-o
+In general, in memory operation should be preferred over creating temporary files.
+If necessary, store them in ``t.TempDir()``. However, in most cases only a
+few methods provided by the Repository are required.
+Then some helper like ``TestTreeMap`` can be used or just create a basic mock yourself.
```
```suggestion
Then some helper like ``data.TestTreeMap`` can be used or just create a basic mock yourself.
```

MichaelEischer commented on doc/090_participating.rst:
```diff
@@ -105,6 +105,129 @@ project <https://dave.cheney.net/2016/03/12/suggestions-for-contributing-to-an-o
+For backends the test suite in ``internal/backend/test`` is mandatory.
+
+For feature flags use ``defer feature.TestSetFlag(t, feature.Flag, feature.DeviceIDForHardlinks, true)()``
+to enable them temporarily in tests.
```
```suggestion
To temporarily enable feature flags in tests, use ``defer feature.TestSetFlag(t, feature.Flag, feature.DeviceIDForHardlinks, true)()``.
```

MichaelEischer commented on doc/090_participating.rst:
`````diff
@@ -105,6 +105,129 @@ project <https://dave.cheney.net/2016/03/12/suggestions-for-contributing-to-an-o
+- ``testRunBackup(t, "", []string{env.testdata}, BackupOptions{}, env.gopts)``: backup all of the standard tree structure
+- ``testListSnapshots(t, env.gopts, <n>)``: check that there are <n> snapshots in the repository
+- ``testRunCheck(t, env.gopts)``: check that the repository is sound and happy
+- the above mentioned ``rtest.OK()``, ``rtest.Equals()``, ``rtest.Assert()````
`````
```suggestion
- the above mentioned ``rtest.OK()``, ``rtest.Equals()``, ``rtest.Assert()`` helpers
```

MichaelEischer reviewed (changes requested):
Please link to this new section from https://github.com/restic/restic/blob/master/CONTRIBUTING.md?plain=1#L203 and https://github.com/restic/restic/blob/master/.github/PULL_REQUEST_TEMPLATE.md?plain=1#L39

wplapper commented on doc/090_participating.rst:
```diff
@@ -105,6 +105,129 @@ project <https://dave.cheney.net/2016/03/12/suggestions-for-contributing-to-an-o
+
+In general, in memory operation should be preferred over creating temporary files.
+If necessary, store them in ``t.TempDir()``. However, in most cases only a
+few methods provided by the Repository are required.
```
done.

wplapper committed:
writing tests - announcement

remove file changelog/unreleased/issue-5252

wplapper committed:
writing tests - implement recommendations

updated the documentation

wplapper committed:
writing tests - creating links 1

create a link to the chapter "writing tests"

wplapper committed:
writing tests - create links 2

create link in TEMPLATE for a new PR to the "writing tests" section in
the documentation.

MichaelEischer reviewed (approved):
LGTM. Thanks!
